text/3628-async-gen-types.md

```diff
@@ -0,0 +1,154 @@
+- Feature Name: `async_gen_types`
+- Start Date: 2024-05-06
+- RFC PR: [rust-lang/rfcs#3628](https://github.com/rust-lang/rfcs/pull/3628)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow the syntax `async T` and `gen T` as types, equivalent to
+`impl Future<Output = T>` and `impl Iterator<Item = T>` respectively. Accept
+them anywhere `impl Trait` can appear.
+
+# Motivation
+[motivation]: #motivation
+
+Users working with asynchronous code may encounter `impl Future<Output = T>`
+types.  Users working with iterators may encounter `impl Iterator<Item = T>`
+types.
+
+These types are long and cumbersome to work with. They may be the first time
+a user will encounter an associated type, and they add verbosity that
+obfuscates the `Output`/`Item` types that people care more about. In
+particular, a function that combines multiple futures or iterators with other
+types requires reading past a lot of syntactic overhead.
+
+Users do not encounter these types when consuming iterators with loops or
+combinators (or in the future producing them with `gen` blocks), or when
+producing or consuming futures using async/await syntax.
+
+The syntax proposed by this RFC provides the same benefits that the current
+`async fn` syntax does (highlighting the future output type), but usable in any
+type rather than only in function return values.
+
+# Explanation
+[explanation]: #explanation
+
+In any context where you can write an `impl Trait` type, you can write
+`async T`, which desugars to `impl Future<Output = T>`:
```

Comment:
this won't work when lifetime needs to be captured right?

```rust
use std::future::Future;

async fn g(a: &str) -> usize {
    a.len()
}

fn h(a: &str) -> impl Future<Output = usize> + '_ {
//                                           ^~~~
    async { a.len() } 
}
```

I suppose #3617 will need to be a prerequisite in any more complex situations assuming you don't want to utter the `impl` word.

```rust
fn i1(a: &str) -> use<'_> async usize {
//                ^~~~~~~
    async { a.len() }
}
fn i2(a: &str) -> use<> async usize {
//                ^~~~~
    async { 0 }
}
```

Reply:
I would expect that in any context where you could write `impl Future<Output = usize> + '_`, you could write `async usize + '_`. You could also use `use`.

Reply:
if we write `async async async async usize + '_` the lifetime attach to the outermost `async`?

| Type | Desugared |
|---|---|
| `async async usize + 'a` | `impl Future<Output = impl Future<Output = usize>> + 'a`<br> (maybe require `async (async usize) + 'a`, like `&(dyn Trait + 'a)`) |
| `async (async usize + 'a)` | `impl Future<Output = impl Future<Output = usize> + 'a>` |
| `async impl Send + 'a` | `impl Future<Output = impl Send> + 'a` |
| `async (impl Send + 'a)` | `impl Future<Output = impl Send + 'a>` |

Since that `+ '_` is part of the TypeParamBounds syntax does this mean `async usize + Send` is allowed :thinking: